feat: Add CSV Export to ankilib

TarekkMA opened the pull request:
## Pull Request template

## Purpose / Description
Adding the functionality to export csv/tsv notes/cards to ankilib. Will implement the GUI for this in another PR

## Fixes
Fixes (partially) #1913

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/Utils.java:
```diff
@@ -107,9 +107,10 @@
     private static final Pattern stylePattern = Pattern.compile("(?si)<style.*?>.*?</style>");
     private static final Pattern scriptPattern = Pattern.compile("(?si)<script.*?>.*?</script>");
-    private static final Pattern tagPattern = Pattern.compile("<.*?>");
+    private static final Pattern tagPattern = Pattern.compile("(?s)<.*?>");
```
Noticed that this regex is missding the dotall modifier found in the upstream

david-allison-1 replied:
Also exists:

https://github.com/ankidroid/Anki-Android/blob/84f1d231ade6caef606dba48a054839acdc516ee/api/src/main/java/com/ichi2/anki/api/Utils.java#L36

david-allison-1 replied:
No changes on the line since 2010, is it possible that we could get a unit test on this one to confirm we're still matching Anki Desktop?

TarekkMA replied:
You mean to ensure java/android acts similarly to python re pacakge ?

david-allison-1 replied:
I'd prefer to go for something higher-level: 

some previously failing code using `tagPattern` now matches the Anki Desktop output.

TarekkMA replied:
oh, okay I will try to figure out some strings that do that

TarekkMA replied:
In normal circumstances `.` in regex matches anything but line breaks, using the `(?s)` modifier will make `.` match anything. So the only case where AnkiDriod would fail if the tag have a line break inside it.

```
will fail on AnkiDriod, but will be mathced on AnkiDesktop
<anki
tag>
```

TarekkMA replied:
I will write a unit test for both pycode and with the AnkiDroid to test this assumption.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/Utils.java:
```diff
@@ -295,25 +296,15 @@ public static String roundedTimeSpan(Context context, long time_s) {
-     * @param s The HTML text to be cleaned.
-     * @return The text without the aforementioned tags.
-     */
-    public static String stripHTMLScriptAndStyleTags(String s) {
```
Not used anywhere else, so I removed it for the sake of simplicity

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/Utils.java:
```diff
@@ -295,10 +296,12 @@ public static String roundedTimeSpan(Context context, long time_s) {
-        Matcher htmlMatcher = tagPattern.matcher(s);
-        s = htmlMatcher.replaceAll("");
+    @NonNull
+    public static String stripHTML(@NonNull String s) {
```
Update `stripHTML` to strip also comments, and refactored the code to look more readable.

david-allison-1 replied:
in `libAnki`, we try to stay as close to Anki Dekstop's code as possible, typically avoiding refactorings unless they're well thought through, and vital.

It's best to go for compatible code, rather than readable code here, as it allows a future developer to compare the two codebases 1:1.

A lot of this has moved to Rust now, and we'll be following in 2.16.x, so some methods may not have Python equivalents

TarekkMA replied:
yeah the code was not 1:1 to the pycode. now it's looking readable + similar to pycode

TarekkMA replied:
![Screen Shot 2021-03-17 at 1 33 11 AM](https://user-images.githubusercontent.com/6633545/111393105-be54bf80-86c0-11eb-9b12-93454e7360dc.png)

TarekkMA replied:
Loving the idea of moving to rust and the sharing the same backend with the desktop, will remove a lot of hassels and let each platform focus on presentation issues, etc

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -74,6 +84,185 @@ public Exporter(Collection col, Long did) {
+
+
+    @NonNull
+    protected String processText(@NonNull String text, boolean includeHTML) {
```
Implementation of this method have `includeHTML ` parameter that is not fond in the upstream.

In the upstream code `includeHTML` is a member field of the base `Exporter` and not all exporters need it, so I couldn't just put it there and force all subclasses to add it to their constructors.

Want to know your opinion about this.

TarekkMA replied:
I think this is the revision I'm using
https://github.com/ankitects/anki/tree/19c217d8a4edc6440bfc768397692404597d5829

david-allison-1 reviewed (commented):
Forgot to submit this, more to go through, review's still on us/me, but looks good for a skim

david-allison-1 reviewed (commented):
I'm going to need to easily compare the Anki Desktop code to review this

* One change has been merged and is causing conflicts (#8256) - can be removed
* Could you provide a source link to the commit in Anki Desktop where each of these comes from
    * Also include this in a comment for the files
* Please include a copyright header for the files you've added
   * This may also require going back to the unit test that you added
   
(No change required yet), be aware that we may need to make platform-specific changes to this (eg: writing to a stream, instead of to a file).

TarekkMA commented:
* Removed the conflicting commit
* Added permlinks to the upstream source I used
* Added copyright to all newly created files

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -74,9 +84,61 @@ public Exporter(Collection col, Long did) {
+        s = s.replaceAll("\\[sound:[^]]+\\]", "");
+        s = Utils.stripHTML(s);
+        s = s.replaceAll("[ \\n\\t]+", " ");
+        s = s.trim(); //s.strip();
```
I believe that Java's `.trim()` works only on spaces and ASCII control characters, whereas `.strip()` also works on unicode-based whitespace.

I'm not certain about this, could you confirm?

TarekkMA replied:
Yes you are right. Python is unicode aware.

`String#strip` is JDK 11 feature that cannot be used. [`StringUtils#strip`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#strip-java.lang.String-) which we can use

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/TextNoteExporter.java:
```diff
@@ -0,0 +1,94 @@
+                final List<String> row = new ArrayList<>();
+
+                if (mIncludeID) {
+                    row.add(String.valueOf(id));
```
`valueOf` can probably be removed - we already converted it to a string

TarekkMA replied:
```java
final String id = cursor.getString(0); // this is guid
```

I know it's confusing, the id variable I get from db is actually guid, but this is how it's named in the upstream.

TarekkMA replied:
Oh right, the upstream did cast it to string so I think I got confused by that.
https://github.com/ankitects/anki/blob/19c217d8a4edc6440bfc768397692404597d5829/pylib/anki/exporting.py#L157

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/TextNoteExporter.java:
```diff
@@ -0,0 +1,94 @@
+                }
+
+                if (mIncludedTags) {
+                    row.add(tags.trim());
```
flagging: strip/trim possible difference

david-allison-1 reviewed (commented):
Looks great (and tests!!)! Thanks

Few comments, nothing major, should be confirmable with another test

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -74,9 +82,60 @@ public Exporter(Collection col, Long did) {
+
+    @NonNull
+    protected String processText(@NonNull String text, boolean includeHTML) {
+        if (!includeHTML) {
```
I wanted to make all variables as final. 

But making it a protected variable and setting them inside CsvExporters constructors could also work.

TarekkMA commented:
@david-allison-1 I have moved `mIncludeHTML` to the base exporter and it's now match the upstream. And I did rebase this breach onto master to solve all conflicts.

Waiting if it's okay to use [`StringUtils#strip`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#strip-java.lang.String-) to solve the mismatched behavior of java trim and python strip.

david-allison-1 commented:
Unsure on this one. I'd be tempted to write it ourselves because it's bringing in a large library/dependency for a small function... but R8 should strip out pretty much everything.

@mikehardy Thoughts?

TarekkMA commented:
It will be easy to implement, we will depend on `Character.isWhitespace` (JDK7) to detect whitespaces.

david-allison-1 commented:
> It will be easy to implement, we will depend on `Character.isWhitespace` (JDK7) to detect whitespaces.

Sounds like the best option to me, thanks!

TarekkMA commented:
+ Added the `strip` function to an already existing `StringUtil` class along with its tests.
+ Added some test for `Exporter#stripHTML` to test that it removes HTML and unicode white spaces.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/TextNoteExporter.java:
```diff
@@ -0,0 +1,93 @@
+                }
+
+                if (mIncludedTags) {
+                    row.add(tags.trim());
```
Still using trim

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/StringUtil.java:
```diff
@@ -37,4 +37,35 @@ public static String trimRight(@Nullable String s) {
+            end--;
+        }
+
+        return string.substring(0, end);
```
Perf nitpicks: 

If `start == 0 && end == .length()` then we made no changes and can return the input string

I think we only need one substring operation, instead of `string.substring(start);`, check if the entire string was a space, and if so, return early.

Otherwise, we have a character in-between, so the start/end are two non-overlapping ranges, and we can use substring to remove both in a single operation.

TarekkMA replied:
fix

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/utils/StringUtilTest.java:
```diff
@@ -46,4 +47,32 @@ public void trimRightDoesNothingOnTrimmedString() {
+        assertEquals("", StringUtil.strip(" \u0020\u2001  "));
+    }
+
+
```
Strip on a non-empty string full of whitespace

TarekkMA replied:
added it to its own test

david-allison-1 reviewed (changes requested):
Only nitpicks, I'd typically be happy to let this go, but Export may be used on a lot of cards, and our users would notice small perf improvements

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -97,6 +98,56 @@ public Exporter(@NonNull Collection col, @NonNull Long did) {
+        s = s.replaceAll("\\[sound:[^]]+\\]", "");
+        s = Utils.stripHTML(s);
+        s = s.replaceAll("[ \\n\\t]+", " ");
+        s = StringUtil.strip(s); //s.strip();
```
```suggestion
        s = StringUtil.strip(s);
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/StringUtil.java:
```diff
@@ -37,4 +37,42 @@ public static String trimRight(@Nullable String s) {
+            return string;
+        }
+
+        if (end == start) {
```
You can move this style of check to above `int end = string.length()`, if a string is whitespace then "start" will have moved to the end

TarekkMA replied:
You are absolutely right,

david-allison-1 reviewed (commented):
Last one I think, thanks!!!

david-allison-1 reviewed (approved):
Fantastic! Looks good to me

Note to maintainer: ensure commits are clean, then rebase. Commit messages contain refs to Anki Codebase and are very useful to keep separate

mikehardy reviewed (approved):
LGTM - I especially love the test support (parameterized even! that's fancy - in a good way - testing, nice)

mikehardy commented:
rebase-merged as requested, all looked clean
